docs/DeveloperGuide.adoc

```diff
@@ -235,6 +235,128 @@ image::CommitActivityDiagram.png[]
 // end::undoredo[]
 
+=== Undo/redo feature
+
+The undo/redo feature lets users undo and redo changes to the data stored in the app, which were
+effected by commands they have executed.
+
+==== Implementation
+
+The main class containing the states of the data in the app is `UndoHistory`. The history is
+represented as a list of states. It includes the following methods:
+
+* `notifyChange()` -- Saves the current state in the history if it is not equal to the previous
+state, as defined by its `equals()` method.
+* `undo()` -- Moves its internal pointer backwards and returns the previous state.
+* `redo()` -- Moves its internal pointer forwards and returns the next state.
+
+`ModelManager` contains an `UndoHistory`, and exposes its functionality via similarly named
+methods. When `undo()` or `redo()` of `ModelManager` is called, it sets its own data to that
+returned by the respective methods of `UndoHistory`.
+
+Given below is an example usage scenario and how the undo/redo mechanism behaves at each step.
+
+Step 1. The user launches the application for the first time. The `UndoHistory` will be initialized with the initial address book state, and the `current` state pointer pointing to that single address book state.
+
+image::UndoRedoState0.png[]
+
+Step 2. The user executes `delete 5` command to delete the 5th person in the address book. After each command, `LogicManager` calls `Model#notifyChange()`, causing the modified state of the data after the `delete 5` command executes to be saved in the `history` list, and the `current` state pointer is shifted to the newly inserted address book state.
+
+image::UndoRedoState1.png[]
+
+Step 3. The user executes `add n/David ...` to add a new person. `LogicManager` calls `Model#notifyChange()` again, causing another modified state to be saved into the `history` list.
+
+image::UndoRedoState2.png[]
+
+[NOTE]
+If a command does not modify the data, when `UndoHistory` checks whether the current data is equal
+to the data in the previous state according to its `equals()` method, it will realise that it is
+indeed equal. It will then not store the state into the `history` list.
+
+Step 4. The user now decides that adding the person was a mistake, and decides to undo that action by executing the `undo` command. The `undo` command will call `Model#undo()`, which will shift the `current` state pointer once to the left, pointing it to the previous state, and restores the data to that state.
+
+image::UndoRedoState3.png[]
+
+[NOTE]
+If the `current` state pointer is at index 0, i.e. pointing to the initial state, then there are no previous states to restore. The `undo` command uses `Model#hasUndo()` to check if this is the case. If so, it will return an error to the user rather than attempting to perform the undo.
+
+The following sequence diagram shows how the undo operation works:
+
+image::UndoSequenceDiagram1.png[]
```

Comment:
The instance name for `UndoCommand` could be more descriptive; calling it `undoCommand` might be more informative.